### ISSUE TITLE
Sorting exception when part of project version contains alpha when others do not

### ISSUE THREAD

**Issue by dougthor42** [rgalanakis/hostthedocs#5]
## Summary
If a project version has an alpha char in the version somewhere, the sorting fails.

## Steps to Reproduce
1. Create a project and start HostTheDocs server.
2. Upload the project with `version = 0.0.1`
3. Upload the project with `version = 0.0.2a`
4. Navigate to HostTheDocs page.
5. See Internal Server Error if running on Apache + WSGI or see exception if running in debug.
6. Rename `hostthedocs/static/docfiles/project/0.0.2a` to `hostthedocs/static/docfiles/project/0.0.2` (removing the alpha from the folder name
7. Navigate to HostTheDocs page.
8. See everything is working fine.

## Comments
Looks like it's an easy enough fix, but there needs to be discussion:
+ Should the alpha be removed and then just sort by int?
+ Should the sorting be changed to alpha sorting?
  + if so, just for that part of the version?
  + or for the entire version string?
+ Should there be warning when trying to upload a file with a version string that will break things?
+ I haven't tested it yet, but based on the source code it looks like if there *always* is an alpha char in that slot of the version, then it will sort fine. For example, these will probably sort fine:
  + `0.1.a`
  + `0.1.b`
  + `0.1.c`

### apache/error.log
```
[Wed Dec 14 13:06:02.437201 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332] [2016-12-14 13:06:02,436] ERROR in app: Exception on / [GET]
[Wed Dec 14 13:06:02.437237 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332] Traceback (most recent call last):
[Wed Dec 14 13:06:02.437251 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1988, in wsgi_app
[Wed Dec 14 13:06:02.437256 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     response = self.full_dispatch_request()
[Wed Dec 14 13:06:02.437259 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1641, in full_dispatch_request
[Wed Dec 14 13:06:02.437262 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     rv = self.handle_user_exception(e)
[Wed Dec 14 13:06:02.437265 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1544, in handle_user_exception
[Wed Dec 14 13:06:02.437269 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     reraise(exc_type, exc_value, tb)
[Wed Dec 14 13:06:02.437272 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
[Wed Dec 14 13:06:02.437275 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     raise value
[Wed Dec 14 13:06:02.437278 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1639, in full_dispatch_request
[Wed Dec 14 13:06:02.437281 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     rv = self.dispatch_request()
[Wed Dec 14 13:06:02.437284 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1625, in dispatch_request
[Wed Dec 14 13:06:02.437287 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     return self.view_functions[rule.endpoint](**req.view_args)
[Wed Dec 14 13:06:02.437290 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/var/www/hostthedocs/hostthedocs/__init__.py", line 37, in home
[Wed Dec 14 13:06:02.437293 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     projects = parse_docfiles(getconfig.docfiles_dir, getconfig.docfiles_link_root)
[Wed Dec 14 13:06:02.437296 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/var/www/hostthedocs/hostthedocs/filekeeper.py", line 42, in parse_docfiles
[Wed Dec 14 13:06:02.437299 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     for f in sorted(os.listdir(docfiles_dir), key=str.lower)]
[Wed Dec 14 13:06:02.437302 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/var/www/hostthedocs/hostthedocs/filekeeper.py", line 42, in <listcomp>
[Wed Dec 14 13:06:02.437305 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     for f in sorted(os.listdir(docfiles_dir), key=str.lower)]
[Wed Dec 14 13:06:02.437308 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]   File "/var/www/hostthedocs/hostthedocs/filekeeper.py", line 28, in _get_proj_dict
[Wed Dec 14 13:06:02.437311 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]     versions.sort(key=sort_by_version)
[Wed Dec 14 13:06:02.437316 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332] TypeError: unorderable types: int() < str()
[Wed Dec 14 13:06:02.437339 2016] [wsgi:error] [pid 20112:tid 139763720873728] [remote 192.168.11.124:32332]
[Wed Dec 14 13:06:02.437412 2016] [wsgi:error] [pid 20112:tid 139763720873728] ERROR:hostthedocs:Exception on / [GET]
[Wed Dec 14 13:06:02.437429 2016] [wsgi:error] [pid 20112:tid 139763720873728] Traceback (most recent call last):
[Wed Dec 14 13:06:02.437432 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1988, in wsgi_app
[Wed Dec 14 13:06:02.437436 2016] [wsgi:error] [pid 20112:tid 139763720873728]     response = self.full_dispatch_request()
[Wed Dec 14 13:06:02.437439 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1641, in full_dispatch_request
[Wed Dec 14 13:06:02.437442 2016] [wsgi:error] [pid 20112:tid 139763720873728]     rv = self.handle_user_exception(e)
[Wed Dec 14 13:06:02.437445 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1544, in handle_user_exception
[Wed Dec 14 13:06:02.437448 2016] [wsgi:error] [pid 20112:tid 139763720873728]     reraise(exc_type, exc_value, tb)
[Wed Dec 14 13:06:02.437450 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
[Wed Dec 14 13:06:02.437453 2016] [wsgi:error] [pid 20112:tid 139763720873728]     raise value
[Wed Dec 14 13:06:02.437456 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1639, in full_dispatch_request
[Wed Dec 14 13:06:02.437459 2016] [wsgi:error] [pid 20112:tid 139763720873728]     rv = self.dispatch_request()
[Wed Dec 14 13:06:02.437462 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/usr/local/pythonenvs/hostthedocs/lib/python3.5/site-packages/flask/app.py", line 1625, in dispatch_request
[Wed Dec 14 13:06:02.437465 2016] [wsgi:error] [pid 20112:tid 139763720873728]     return self.view_functions[rule.endpoint](**req.view_args)
[Wed Dec 14 13:06:02.437468 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/var/www/hostthedocs/hostthedocs/__init__.py", line 37, in home
[Wed Dec 14 13:06:02.437470 2016] [wsgi:error] [pid 20112:tid 139763720873728]     projects = parse_docfiles(getconfig.docfiles_dir, getconfig.docfiles_link_root)
[Wed Dec 14 13:06:02.437473 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/var/www/hostthedocs/hostthedocs/filekeeper.py", line 42, in parse_docfiles
[Wed Dec 14 13:06:02.437476 2016] [wsgi:error] [pid 20112:tid 139763720873728]     for f in sorted(os.listdir(docfiles_dir), key=str.lower)]
[Wed Dec 14 13:06:02.437479 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/var/www/hostthedocs/hostthedocs/filekeeper.py", line 42, in <listcomp>
[Wed Dec 14 13:06:02.437482 2016] [wsgi:error] [pid 20112:tid 139763720873728]     for f in sorted(os.listdir(docfiles_dir), key=str.lower)]
[Wed Dec 14 13:06:02.437484 2016] [wsgi:error] [pid 20112:tid 139763720873728]   File "/var/www/hostthedocs/hostthedocs/filekeeper.py", line 28, in _get_proj_dict
[Wed Dec 14 13:06:02.437487 2016] [wsgi:error] [pid 20112:tid 139763720873728]     versions.sort(key=sort_by_version)
[Wed Dec 14 13:06:02.437491 2016] [wsgi:error] [pid 20112:tid 139763720873728] TypeError: unorderable types: int() < str()
[Wed Dec 14 13:06:02.437498 2016] [wsgi:error] [pid 20112:tid 139763720873728]
```

**Comment by rgalanakis** [comment]
Thanks for such a detailed report! I think this is ultimately an issue with Python 2/3 compat. `int` and `str` are orderable in Python 2, but not Python 3- I think ints always sort "lower" than chars. The `sort_by_version` function depends on ints and strs being orderable and it ends up sorting quite nicely actually :)

The easiest thing to do would probably be to use `cmp` instead of `key` in the call to `versions.sort`. The function provided to `cmp` could do the custom "if `a` is an int and `b` is a string, then `a` is always less than `b`" logic (and all permutations), rather than depend on Python 2's built-in comparison logic. There are some other options but I think that's the simplest.

I'd happily merge a PR if you want to put one up; otherwise, I can take a stab, but I'd much prefer a contribution! Thanks again for the report.

**Comment by allebacco** [comment]
A possible solution to the sorting problem could be the use of natsort package (http://natsort.readthedocs.io/en/stable/) that is able to sort version numbers with both numbers and chars.

**Comment by rgalanakis** [comment]
Fixed in #9
